src/talk/renderer/UserStatus/useUserStatusHeartbeat.js

```diff
@@ -0,0 +1,90 @@
+/*
+ * @copyright Copyright (c) 2024 Grigorii Shartsev <me@shgk.me>
+ *
+ * @author Grigorii Shartsev <me@shgk.me>
+ *
+ * @license AGPL-3.0-or-later
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { onBeforeUnmount, ref } from 'vue'
+import { useUserStatusStore } from './userStatus.store.js'
+
+/** How often to update the user status */
+const USER_STATUS_UPDATE_INTERVAL = 5 * 60 * 1000 // 5 minutes
+
+/** How log user is considered active */
+const USER_STATUS_ACTIVE_TIMEOUT = 2 * 60 * 1000 // 2 minutes
+
+/**
+ * Hook to update the user status in the background
+ * @return {{ isActive: import('vue').Ref<boolean> }}
+ */
+export function useUserStatusHeartbeat() {
+	const userStatusStore = useUserStatusStore()
+
+	let heartbeatInterval
+	const restartHeartbeat = () => {
+		if (heartbeatInterval) {
+			clearInterval(heartbeatInterval)
+		}
+		heartbeatInterval = setInterval(async () => {
+			try {
+				await userStatusStore.updateUserStatusWithHeartbeat(!isActive.value)
+				// userStatusStore.setUserStatus(updatedStatus)
+			} catch {}
+		}, USER_STATUS_UPDATE_INTERVAL)
+	}
+
+	restartHeartbeat()
+
+	const isActive = ref(true)
+
+	let isActiveTimeout = null
+	const markActive = () => {
+		if (!isActive.value) {
+			// Become active
+			isActive.value = true
+			restartHeartbeat()
+		}
+
+		// Reset active after some time of inactivity
+		clearTimeout(isActiveTimeout)
+		isActiveTimeout = setTimeout(() => {
+			isActive.value = false
+		}, USER_STATUS_ACTIVE_TIMEOUT)
+	}
+
+	const handleVisibilityChange = () => {
+		if (document.hidden) {
+			isActive.value = false
+		} else {
+			markActive()
+		}
+	}
+
+	window.addEventListener('mousemove', markActive, { passive: true, capture: true })
+	window.addEventListener('keydown', markActive, { passive: true, capture: true })
+	document.addEventListener('visibilitychange', handleVisibilityChange, { passive: true })
```

Comment:
maybe `touchstart` too ?

Reply:
Agree and added some more events.

Also, moved checking for active (idle) to a separate composable.